fbink.c

```diff
@@ -5645,70 +5636,34 @@ static __attribute__((cold)) int
 	}
 	otInit = true;
 
-	// Open font from given path, and load into buffer
-	FILE* f                      = fopen(filename, "r" STDIO_CLOEXEC);
-	unsigned char* restrict data = NULL;
-	if (!f) {
-		PFWARN("fopen: %m");
-		otInit = false;
-		return ERRCODE(EXIT_FAILURE);
-	} else {
-		const int   fd = fileno(f);
-		struct stat st;
-		if (fstat(fd, &st) == -1) {
-			PFWARN("fstat: %m");
-			fclose(f);
-			otInit = false;
-			return ERRCODE(EXIT_FAILURE);
-		}
-		data = calloc((size_t) st.st_size, sizeof(*data));
-		if (!data) {
-			PFWARN("Error allocating font data buffer: %m");
-			fclose(f);
-			otInit = false;
-			return ERRCODE(EXIT_FAILURE);
-		}
-		if (fread(data, 1U, (size_t) st.st_size, f) < (size_t) st.st_size || ferror(f) != 0) {
-			free(data);
-			fclose(f);
-			otInit = false;
-			WARN("Error reading font file `%s`", filename);
-			return ERRCODE(EXIT_FAILURE);
-		}
-		fclose(f);
-	}
-	stbtt_fontinfo* font_info = calloc(1U, sizeof(stbtt_fontinfo));
+    stbtt_fontinfo* font_info = calloc(1U, sizeof(stbtt_fontinfo));
 	if (!font_info) {
 		PFWARN("Error allocating stbtt_fontinfo struct: %m");
-		free(data);
 		return ERRCODE(EXIT_FAILURE);
 	}
 	// First, check if we can actually find a recognizable font format in the data...
 	const int fontcount = stbtt_GetNumberOfFonts(data);
 	if (fontcount == 0) {
-		free(data);
 		free(font_info);
-		WARN("File `%s` doesn't appear to be a valid or supported font", filename);
+		WARN("File doesn't appear to be a valid or supported font");
 		return ERRCODE(EXIT_FAILURE);
 	} else if (fontcount > 1) {
-		LOG("Font file `%s` appears to be a font collection containing %d fonts, but we'll only use the first one!",
-		    filename,
+		LOG("Font file appears to be a font collection containing %d fonts, but we'll only use the first one!",
 		    fontcount);
 	}
 	// Then, get the offset to the first font
 	const int fontoffset = stbtt_GetFontOffsetForIndex(data, 0);
 	if (fontoffset == -1) {
-		free(data);
```

Comment:
In the same vein, this needs to be `free(font_info->data);`

Which impliers that we take ownership of the pointer, which might be a problem for your usecase (e.g., pointer to static content), so this may need to be more complex to add a way to say "nope, this data is immutable, don't ever try to free it).

Reply:
Weeelll, except `font_info->data` only exists *after* `stbtt_InitFont` ;p.

The point about ownership still stands, though ;p.